### ISSUE TITLE
upgrade '--all' and '--accept-package-agreements' flags should work together

### ISSUE THREAD

**Issue by MaZe3D** [microsoft/winget-cli#2680]
### Description of the new feature / enhancement

I tried to use the command `winget upgrade --all` command but after the package repos were successfully downloaded, it aborted because the package-agreements werent accepted. But I was not asked either if I tried to upgrade all. So I tried to append the `--accept-package-agreements` flag, but assumingly this won't work in combination with `--all`.

The annoying part is that it abborts the whole proccess of upgrading all packages and wont continue. So I have to manually upgrade all packages with included package-agreements in order to accept it, and then use --all to upgrade all others.

Without using both flags in combination it makes the whole purpurse of --all obsolete because it wont run no matter if you want or wont want to accept the package agreements

### Proposed technical implementation details

I want the command `winget upgrade --all --accept-package-agreements` to work and upgrade all packages and automatically accept the agreemets, if nessecary.

**Comment by awecomp** [comment]
Just had the same issue, one program wouldn't upgrade with "Package agreements were not agreed to. Operation cancelled."
"The arguments provided can only be used with a query." with "winget upgrade --accept-package-agreements --all"
Ended up just uninstalling the software that was causing issues and upgraded the rest.

**Comment by eabase** [comment]
Yeah, this is not good for automation. Also make sure that `--accept-source-agreements` is also compatible.

**Comment by joshuatrevissmith** [comment]
I have this problem when I install using Windows Store source, because I am prompted to accept the agreement upon install and every update. 

`winget --all` will fail with:
```
Package agreements were not agreed to. Operation cancelled.
```

The workaround has been to first use winget or Store app to update the store sourced package(s), then `winget --all` will update the remaining winget sourced packages that I have.

**Comment by gellenburg** [comment]
Just adding that this definitely needs to be fixed:

`winget upgrade --all --accept-package-agreements --accept-source-agreements`

What's the point in having an `upgrade --all` flag if the flag won't even work?

If I can do an `rpm -Uvh` or `apt-get -y upgrade` on Linux, I should be able to do an `winget upgrade --all --accept-package-agreements --accept-source-agreements` as well.

**Comment by soulhax** [comment]
I will bump this as well. This must be fixed ASAP.

**Comment by eabase** [comment]
This is a rather serious UX issue, with what appears to be an easy fix. 
Would be great to get some ETA update/feedback from the MS Team.

**Comment by rootomatic** [comment]
as a workaround you're able to enter the Y automatically by piping it in powershell

ECHO Y | winget.exe" upgrade --all --silent

**Comment by qJake** [comment]
> ECHO Y | winget upgrade --all --silent

Didn't work for me.

This should really be fixed. Either allow `--all --accept-package-agreements --accept-source-agreements`, or add a separate command like `winget agreements --accept-all` that flags any upgrades with "accepted" license terms so that the upgrade can proceed.

**Comment by kriswme2** [comment]
Supplying a dummy package search query works for me
```winget upgrade --accept-package-agreements --all <any dummy string>```

I've just upgraded all the packages with ```winget upgrade --accept-package-agreements --all 1```
![image](https://user-images.githubusercontent.com/9313114/211703562-30e74f13-4845-4e00-abbf-f34441183b9c.png)

**Comment by rootomatic** [comment]
> > ECHO Y | winget upgrade --all --silent
> 
> Didn't work for me.
> 
> This should really be fixed. Either allow `--all --accept-package-agreements --accept-source-agreements`, or add a separate command like `winget agreements --accept-all` that flags any upgrades with "accepted" license terms so that the upgrade can proceed.

Looks like I missed a little off the command I used. It's actually 
`ECHO Y | cmd /c winget upgrade --all --silent --include-unknown`

**Comment by rootomatic** [comment]
> > ECHO Y | winget upgrade --all --silent
> 
> Didn't work for me.
> 
> This should really be fixed. Either allow `--all --accept-package-agreements --accept-source-agreements`, or add a separate command like `winget agreements --accept-all` that flags any upgrades with "accepted" license terms so that the upgrade can proceed.

Yeah looks like I missed a couple a bit. In powershell this works for me
`ECHO Y | cmd /c winget upgrade --all --silent`

**Comment by eabase** [comment]
Does this merge mean it's been fixed?

**Comment by artem-korolev** [comment]
> Does this merge mean it's been fixed?

seems it has been released in preview version - https://github.com/microsoft/winget-cli/releases/tag/v1.5.101-preview

"Copy install behavior flags on upgrade --all by @florelis in https://github.com/microsoft/winget-cli/pull/2794"

waiting for release version. don't wanna mess with pre-release version personally

**Comment by denelon** [comment]
PR Merged is just an indication that code has been merged in "main" branch (assuming that's the target).
Releases have links to included PRs (in this case it's in a preview release).
If the work has been merged for the scope indicated in the issue, then the issue can be closed. 
A closed issue is not an indication that the "fix" or "feature" is in a release yet.

For this issue, additional work to be able to close it is in progress with:
* https://github.com/microsoft/winget-cli/pull/2862

Once that work is complete and verified, this issue will be closed.

It will not be available (other than in a developer's individual build) until we cut our next "preview" release after it has been merged. 

When we cut the next "stable" release after the work is in a "preview" release, then it will be GA (Generally Available).

This is being tracked in our 1.5 milestone.

**Comment by MatheusRodri** [comment]
I'm trying to use the command `winget upgrade --all --accept-package-agreements --accept-source-agreements` but  appear another error: `The arguments provided can only be used with a query.` How do I resolve this?

**Comment by denelon** [comment]
We've validated the referenced PR addresses these concerns.

This fix will be available in our next preview release after 1.5-preview(1).

**Comment by Salman7236** [comment]
winget upgrade -u -h -r --accept-source-agreements --accept-package-agreements

Is there something wrong with the above command? it gives me this error:
"The arguments provided can only be used with a query."

If I remove the last two agreement commands I get this error:
"Package agreements were not agreed to. Operation cancelled."

**Comment by denelon** [comment]
No package has been specified for the upgrade. If you want to upgrade all packages add "--all".

**Comment by brianhm** [comment]
Can confirm that the latest preview release does honor the switches.  This was my Powershell:

Add-AppxPackage c:\temp\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle
winget upgrade -h --all --accept-package-agreements --accept-source-agreements

I do have a question though.  I thought WinGet was supposed to update all Store Apps with the -all switch?  When I open the Microsoft Store app on Windows 11, it shows more stuff.  Example is on a newly upgraded machine from Windows 10, the Snipping Tool has an updated version, but the winget command above does not find it and update it.  I have to do that manually via the store.  I've tried running the command with Administrator and User context.  Is there a reason for this?  Just some examples in my attached snip.
![image](https://user-images.githubusercontent.com/103964858/225960823-37910b3b-94ba-4bee-8637-2514de1aea98.png)

**Comment by Salman7236** [comment]
> No package has been specified for the upgrade. If you want to upgrade all packages add "--all".

The help menu shows that -r and --all are the same. But I also tried it with --all and got the same error.

**Comment by denelon** [comment]
Try `winget upgrade -uhr --accept-source-agreements --accept-package-agreements`

Any Microsoft Store packages installed with WinGet will show version "Unknown", and Microsoft Store packages not installed with WinGet will not be included. We have future work to address this.

The source and package agreements are only necessary for the Microsoft Store source.

**Comment by brianhm** [comment]
> 

So how do we deal with this in the meantime?  What is the scripted way to update all Store apps on a computer?  We want to do this on a recurring basis with our monthly application patching process like we do with other apps.  All of the forums I read say "use winget for Windows 11" but this appears to not yet be supported.

**Comment by brianhm** [comment]
And btw, this -uhr switch appears to update our Microsoft Office 365 Apps for Enterprise, which we DO NOT want.  I find this bizarre that winget would somehow manage this when we deploy it via ODT.  Do you also have a command line for exclusions in the works?

**Comment by awecomp** [comment]
> > 
> 
> So how do we deal with this in the meantime? What is the scripted way to update all Store apps on a computer? We want to do this on a recurring basis with our monthly application patching process like we do with other apps. All of the forums I read say "use winget for Windows 11" but this appears to not yet be supported.

I was wondering about this, too. Has been an issue especially with per-user apps and not being able to upgrade the Store apps without having to open the Store manually. Hopefully this could be rolled up into Winget maybe so they can get upgraded at the same time - handy for auto-maintenance/scripting/new computer builds/imaging etc!

**Comment by alexdatsko** [comment]
Bump. This needs to be fixed! A real package manager should require no user input when directed to do so.  There should be UAC bypasses for these updates as well as this is core security SOP..

**Comment by MaZe3D** [comment]
I am not agreeing with the UAC bypass tough. I run my command prompt as admin if necessary. That will bypass it but you need to keep in mind that it will run every installer as admin.

**Comment by qJake** [comment]
That gets tricky... some installers change the location depending on how they are run (e.g. non-elevated goes to `%localappdata%` and elevated goes to `%programfiles%`). So if you elevate your terminal, you may see a difference in behavior (possibly unwanted), but if you don't elevate it, your command may not succeed.

**Comment by brianhm** [comment]
Just to bring this topic back around.  The original request was for the cli to update all installed store apps when it currently does not.

**Comment by denelon** [comment]
@brinham I'd suggest creating a new issue, but I can share:

We need to get the version information from the Microsoft Store REST source so we know when there are available updates to those packages. Currently, we only get "Unknown" version for UWP and Win32 packages. Some work has been done to get version for Win32 packages, but not in all the endpoints, and sadly the one we need for version comparison still renders "Unknown". We're working with them to unblock the scenario.

**Comment by rogacz1988** [comment]
Any updates? I still get that error.

**Comment by NireBryce** [comment]
still get this error, still seemingly no way to accept all the agreements.

**Comment by jaliyaudagedara** [comment]
For some reason, I still need to explicitly accept *.

If you are PowerShell, came up with this workaround, 

Edit PowerShell profile
```
code $PROFILE
# or 
notepad $PROFILE
```
Add the following function,
```
# Short command to upgrade all the installed packages
function Winget-Upgrade-All() {    
    winget upgrade --all --accept-source-agreements --accept-package-agreements
}
```
Execute from PowerShell
```
Winget-Upgrade-All
```
![image](https://github.com/microsoft/winget-cli/assets/5653381/6bc828b0-db69-4a6a-a193-f60a97acdd3b)

**Comment by rbarbrow** [comment]
winget upgrade --all --accept-source-agreements --accept-package-agreements   is not intuitive at all

Winget upgrade --all should at least upgrade the ones that can and just ignore the ones that can't rather than failing the entire task

**Comment by shamimcloud** [comment]
This command work:

Winget upgrade --all --accept-package-agreements

**Comment by rbarbrow** [comment]
--all should be the implication of acceptance   there should be a flag to reject agreements 

99.99999% of people are going to accept due to the fact that they are upgrading software already on their machines the burden should be on the minority to flag not for the majority 

it makes the software harder to use and isn't immediately obvious to non experts

additionally several tutorials already exist in the wild that give the --all command

**Comment by michaelog88** [comment]
I have tried this cmd "Winget upgrade --all --accept-package-agreements"  it's not still working for me with error: Package agreements were not agreed to. Operation cancelled

**Comment by qJake** [comment]
> I have tried this cmd "Winget upgrade --all --accept-package-agreements" it's not still working for me with error: Package agreements were not agreed to. Operation cancelled

Per comments earlier in this thread, have you tried `winget upgrade --all --accept-source-agreements --accept-package-agreements` ?

**Comment by eabase** [comment]
With 2 years of highly functional AI's & LLM's and 4 years of MS *"looking at"*, I rather rudely expected that this would have been fixed. 

I suggest MS should add another CLI switch: 
`--obscure-long-and-undocumented-that-will-be-ignored-unless-randomly-combined-with-ten-other-obscure-switches`

**Comment by denelon** [comment]
Running `winget upgrade --all` is designed up upgrade all installed "packages" with correlated manifests in all configured sources without an explicit targeting requirement. WinGet ships with two sources enabled without an explicit targeting requirement by default. Those two sources are "winget" and "msstore".

The "msstore" source has an agreement that must be accepted to prior to use.
Some "packages" also have agreements that must be accepted prior to installation or upgrade.

These agreements may be changed by the publisher periodically. That places the burden on the "user" to determine if they agree to the current agreements at the time they run the command.

WinGet does list all the "options" and their definitions in help, and at the end of WinGet help, a link is provided with a link to the content at Microsoft Learn.

I've included the key three options from a subset of the output from `winget upgrade --help` below.

| option | description |
| ------------- | ------------- |
| -r,--recurse,--all | Upgrade all installed packages to latest if available |
|  --accept-package-agreements | Accept all license agreements for packages |
|  --accept-source-agreements | Accept all source agreements during source operations |

The last line of output from the command also includes
> More help can be found at: https://aka.ms/winget-command-upgrade

If the user running the command has already accepted the agreements for the source on the device (and the agreements haven't changed), then the argument for "--accept-source-agreements" isn't necessary. In some cases, that causes confusion, so I added it for clarity.

We've been required to ensure agreements are explicitly agreed to which is why using the option "--all" isn't sufficient by itself in some cases to have all packages upgraded in all cases.

**Comment by eabase** [comment]
@denelon 

> *We've been required to ensure agreements are explicitly agreed to which is why using the option "--all" isn't sufficient by itself in some cases to have all packages upgraded in all cases.*

Thanks for clear explanation, BUT... MS doesn't really give AF about any agreements because **I have certainly never agreed to MS sending my clipboard data (that often contain remote server passwords and authentication tokens) to random unknown 3rd party servers**. MS can't even be bothered to name these services in a transparent way. So yeah, it's very hard for us users to take any MS agreements serious. Just skip all the nonsense and make your OS simple and friendly to use, and everybody will be happy, including MS.

**Comment by nemchik** [comment]
> [@denelon](https://github.com/denelon)
> 
> > _We've been required to ensure agreements are explicitly agreed to which is why using the option "--all" isn't sufficient by itself in some cases to have all packages upgraded in all cases._
> 
> Thanks for clear explanation, BUT... MS doesn't really give AF about any agreements because **I have certainly never agreed to MS sending my clipboard data (that often contain remote server passwords and authentication tokens) to random unknown 3rd party servers**. MS can't even be bothered to name these services in a transparent way. So yeah, it's very hard for us users to take any MS agreements serious. Just skip all the nonsense and make your OS simple and friendly to use, and everybody will be happy, including MS.

Not saying I disagree, but the way you're approaching the conversation isn't very likely to get a positive response or results from the devs that could actually make it happen.

**Comment by eabase** [comment]
@nemchik

> *...but the way you're approaching the conversation isn't very likely to get a positive response or results from the devs that could actually make it happen.*

At this point, 4 years later and hidden away in an already closed issue, it is clear that they will never "*make it happen*".  
So I really don't care. When my current HW expires, I will never again use anything from MS.

**Comment by mikebeeboo** [comment]
I had to run it in an Administrator PowerShell session and it ran flawlessly fwiw...

`winget upgrade --all --accept-package-agreements --accept-source-agreements`

**Comment by TheBeardofKnowledge** [comment]
I think what everyone here is looking for is this:
winget update --all --include-unknown --accept-source-agreements --accept-package-agreements --silent --verbose
